packages/nimble-components/src/utilities/models/tracker.ts

```diff
@@ -0,0 +1,62 @@
+type ObjectFromList<T extends readonly string[], V = string> = {
+    [K in T extends readonly (infer U)[] ? U : never]: V;
+};
+
+/**
+ * Generic Tracker which sets or resets flags and utilities
+ */
+export abstract class Tracker<WhimsList extends readonly string[]> {
+    private _whims: ObjectFromList<WhimsList, boolean>;
+
+    public constructor(whimsList: WhimsList) {
+        type Whims = typeof this._whims;
+        this._whims = {} as Whims;
+        this._whims = whimsList.reduce<Whims>((r, key): Whims => {
+            return {
+                ...r,
+                [key]: false
+            };
+        }, this._whims);
+    }
+
+    public get whims(): ObjectFromList<WhimsList, boolean> {
+        return this._whims;
+    }
+
+    public track(key: keyof ObjectFromList<WhimsList, boolean>): void {
```

Comment:
Perhaps all of these methods should be `protected`? This would place a slight burden on the derived implementations where they would need to define their own wrapper APIs around these, but that would address some of the semantic concerns.

Reply:
I think that will limit the usage of the tracker class on its own. Maybe using it as a private property of the validator could work?

Reply:
Oh sorry, I had it in my head that `Tracker` was `abstract` since the derived `UpdateTracker` was `abstract`.